pepper-apis/src/main/java/org/broadinstitute/ddp/service/WorkflowService.java

```diff
@@ -75,4 +95,70 @@ public WorkflowResponse buildStateResponse(Handle handle, String userGuid, Workf
         }
         return WorkflowResponse.unknown();
     }
+
+    private void createActivityInstanceIfMissing(
+            Handle handle,
+            WorkflowState fromState,
+            WorkflowState nextState,
+            String operatorGuid,
+            String userGuid,
+            long studyId
+    ) {
+        if (nextState.getType() != StateType.ACTIVITY) {
+            return;
+        }
+        ActivityState activityState = (ActivityState) nextState;
+        ActivityDto activityDto = handle.attach(JdbiActivity.class).queryActivityById(activityState.getActivityId());
+        if (activityDto.getMaxInstancesPerUser() != null && activityDto.getMaxInstancesPerUser() == 0) {
+            return;
+        }
+        String instanceGuid = handle.attach(JdbiActivityInstance.class)
+                .findLatestInstanceGuidByUserGuidAndActivityId(userGuid, activityState.getActivityId())
+                .orElse(null);
+        boolean activityInstanceIsMissing = instanceGuid == null;
+        if (activityInstanceIsMissing) {
+            ActivityInstanceDto instanceDto = handle.attach(ActivityInstanceDao.class)
+                    .insertInstance(activityState.getActivityId(), operatorGuid, userGuid, InstanceStatusType.CREATED, false);
+            LOG.info("Created start activity instance with guid '{}' for user guid '{}' using operator guid '{}' and activity id {}",
+                    instanceDto.getGuid(), userGuid, operatorGuid, activityState.getActivityId());
+            processActionsForActivityCreationSignal(
+                    handle,
+                    operatorGuid,
+                    instanceDto.getParticipantId(),
+                    userGuid,
+                    instanceDto.getId(),
+                    activityState.getActivityId(),
+                    studyId
+            );
+            LOG.info(
+                    "Processed actions for the activity instance {} of activity {} triggered by the creation signal",
+                    instanceDto.getGuid(),
+                    activityState.getActivityId()
+            );
+        } else {
+            LOG.info("User guid '{}' already has activity instance with guid '{}', nothing to create", userGuid, instanceGuid);
+        }
+    }
+
+    private void processActionsForActivityCreationSignal(
+            Handle handle,
+            String operatorGuid,
+            long participantId,
+            String participantGuid,
+            long instanceId,
+            long activityId,
+            long studyId
+    ) {
+        long operatorId = handle.attach(JdbiUser.class).getUserIdByGuid(operatorGuid);
```

Comment:
Observation: the DAO method returns `Optional` while the column in `study_activity.study_id` is `NOT NULL`. This means that `Optional<Long>` can be replaced with just `long`. Let's clean it up later in another PR.